packages/perspective-viewer-d3fc/src/js/legend/legend.js

```diff
@@ -8,48 +8,119 @@
-            legend.labels(options => {
-                const parts = options.domain[options.i].split("|");
-                return parts.slice(0, parts.length - 1).join("|");
-            });
+        if (doesLegendScroll) {
+            const reloadLegend = index => createLegend(container, settings, colour, doesLegendScroll, groupSize, index);
+            legendScrolling(container, settings, numberOfGroups, reloadLegend);
+        }
+        createLegend(container, settings, colour, doesLegendScroll, groupSize, settings.legendGroupIndex);
+    }
+}
+
+function legendScrolling(container, settings, numberOfGroups, reloadLegend) {
+    if (!settings.legendGroupIndex) settings.legendGroupIndex = 0;
+    const legendControls = getLegendControls(container);
+
+    const setPageText = () => legendControls.select("#page-text").text(`${settings.legendGroupIndex + 1}/${numberOfGroups}`);
+    setPageText();
+    const upArrow = legendControls.select("#up-arrow");
+    const downArrow = legendControls.select("#down-arrow");
+
+    upArrow.on("click", function() {
+        if (settings.legendGroupIndex > 0) {
+            settings.legendGroupIndex--;
+            legendPage(downArrow);
         }
+        deactivateArrow(upArrow, 0);
+    });
+    deactivateArrow(upArrow, 0);
 
-        let legendSelection = container.select("svg.legend");
-        if (legendSelection.size() === 0) {
-            legendSelection = container.append("svg");
+    downArrow.on("click", function() {
+        if (settings.legendGroupIndex < numberOfGroups - 1) {
+            settings.legendGroupIndex++;
+            legendPage(upArrow);
         }
+        deactivateArrow(downArrow, numberOfGroups - 1);
+    });
+    if (settings.legendGroupIndex < numberOfGroups - 1) activateArrow(downArrow);
+
+    function legendPage(arrow) {
+        reloadLegend(settings.legendGroupIndex);
+        setPageText();
+        activateArrow(arrow);
+    }
 
-        // render the legend
-        legendSelection
-            .attr("class", "legend")
-            .style("z-index", "2")
-            .call(legend)
-            .select("g.legendCells")
-            .attr("transform", "translate(20,20)")
-            .selectAll("g.cell")
-            .classed("hidden", data => settings.hideKeys && settings.hideKeys.includes(data));
+    function activateArrow(arrow) {
+        arrow.style("color", "rgb(63, 127, 253)").style("cursor", "pointer");
     }
+
+    function deactivateArrow(arrow, threshold) {
+        if (settings.legendGroupIndex === threshold) arrow.style("color", null).style("cursor", "default");
+    }
+}
+
+function getLegendControls(container) {
+    let legendControls = container.select(".legend-controls");
+    if (legendControls.size() === 0) {
+        legendControls = container
+            .append("div")
+            .attr("class", "legend-controls")
+            .html(legendControlsTemplate);
+    }
+    return legendControls;
+}

```

Comment:
This kind of thing appears a few times, maybe it should be moved into its own function.

Reply:
I was thinking of having a function (possible in utils) where you would pass a selector and a callback for if the element doesn't exist.